WooCommerce/WooCommerceTests/Tools/InfiniteScroll/ScrollWatcherTests.swift

```diff
@@ -0,0 +1,34 @@
+import XCTest
+@testable import WooCommerce
+
+final class ScrollWatcherTests: XCTestCase {
+    private var cancellable: ObservationToken?
+
+    func test_scroll_trigger_is_emitted_only_after_scrolling_to_above_threshold() {
+        // Arrange
+        let frame = CGRect(origin: .zero, size: CGSize(width: 320, height: 100))
+        let tableView = UITableView(frame: frame, style: .grouped)
+        let contentSize = CGSize(width: 320, height: 1100)
+        tableView.contentSize = contentSize
+        let threshold = 0.8
+        let scrollWatcher = ScrollWatcher(scrollPositionThreshold: threshold)
+        scrollWatcher.startObservingScrollPosition(tableView: tableView)
+        var triggeredScrollPositions = [Double]()
+        cancellable = scrollWatcher.scrollTrigger.subscribe { scrollPosition in
+            triggeredScrollPositions.append(scrollPosition)
+        }
+
+        // Action
+        // Scrolls to 50%
+        tableView.scrollRectToVisible(CGRect(origin: CGPoint(x: 0, y: 500), size: frame.size), animated: false)
+        // Scrolls to 70%
+        tableView.scrollRectToVisible(CGRect(origin: CGPoint(x: 0, y: 700), size: frame.size), animated: false)
+        // Scrolls to 80%
+        tableView.scrollRectToVisible(CGRect(origin: CGPoint(x: 0, y: 800), size: frame.size), animated: false)
+        // Scrolls to 100%
+        tableView.scrollRectToVisible(CGRect(origin: CGPoint(x: 0, y: 1000), size: frame.size), animated: false)
```

Comment:
I wonder if we should consider scrolling down versus scrolling up. I added this after this line:

```swift
        tableView.scrollRectToVisible(CGRect(origin: CGPoint(x: 0, y: 900), size: frame.size), animated: false)
        tableView.scrollRectToVisible(CGRect(origin: CGPoint(x: 0, y: 800), size: frame.size), animated: false)
```

The value of `triggeredScrollPositions` is then:

```
[0.8, 1.0, 0.9, 0.8]
```

I think this means we're also triggering events for scrolling up. Do you think this is something to consider?

Reply:
great suggestion! I added a check to only emit a value if it's higher than the previous one in https://github.com/woocommerce/woocommerce-ios/pull/2747/commits/ef9ace08fe4fc1fab9ad8ba04ab0dbb546ede84f